### ISSUE TITLE
cannot build 21.5.0 using any recent rustc version I can lay my hands on.

### ISSUE THREAD

**Issue by drook** [getsentry/snuba#2119]
Env:
-----
Sentry on premises, containerless
rustc/cargo 1.49.0
rustc/cargo 1.54.0
clang version 11.0.1 (git@github.com:llvm/llvm-project.git llvmorg-11.0.1-0-g43ff75f2c3fe)
python 3.9

Issue:
-----
Trying to follow the upgrade path 20.8.0 (mine version) -> 21.5.0 -> recent 21.x

Snuba 21.5.0 compilation fails with the log below (very similar to the https://github.com/getsentry/relay/issues/915 , caused by upstream https://github.com/luser/rust-minidump/issues/100 , fixed long ago there).

Prior to the log meditation - can I ask (yup, this is more a forum question, but for those who will probably find this afterwards - better to accumulate all the knowledge in one place) -
1) do I have to run snuba version in sync with sentry (in general) ? I've tried to build 21.8.0 and it builds just fine.
2) furthermore, do I even need snuba/relay to follow the upgrade path - do I understand correctly that I only need sentry (which I have built) to run it's 21.5.0 migration sequence (**_sentry upgrade_** launch) against clickhouse/postgresql, then proceed to the recent 21.x release ?

Build log:
-----
````
[...]
   Compiling smallvec v1.4.0
   Compiling chrono v0.4.11
   Compiling uuid v0.8.1
   Compiling serde_yaml v0.8.11
   Compiling erased-serde v0.3.11
   Compiling sentry-release-parser v0.6.0
   Compiling globset v0.4.5
   Compiling minidump-common v0.2.0 (https://github.com/luser/rust-minidump?rev=4d95707b20bf15fb37de5f6d7ba7d7e0f8ab4afd#4d95707b)
   Compiling unicode-normalization v0.1.12
   Compiling parking_lot_core v0.7.1
   Compiling debugid v0.7.2
error[E0277]: the trait bound `[u16; 260]: TryFromCtx<'_, _>` is not satisfied
    --> /home/emz/.cargo/git/checkouts/rust-minidump-836c4af26fbfd1e5/4d95707/minidump-common/src/format.rs:1322:25
     |
1322 |           #[derive(Clone, Pread, SizeWith)]
     |                           ^^^^^ the trait `TryFromCtx<'_, _>` is not implemented for `[u16; 260]`
...
1331 | / multi_structs! {
1332 | |     /// Miscellaneous process information
1333 | |     ///
1334 | |     /// This struct matches the [Microsoft struct][msdn] of the same name.
...    |
1376 | |     }
1377 | | }
     | |_- in this macro invocation
     |
     = help: the following implementations were found:
               <&'a [u8] as TryFromCtx<'a, usize>>
     = note: this error originates in the derive macro `Pread` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `[u16; 40]: TryFromCtx<'_, _>` is not satisfied
    --> /home/emz/.cargo/git/checkouts/rust-minidump-836c4af26fbfd1e5/4d95707/minidump-common/src/format.rs:1322:25
     |
1322 |           #[derive(Clone, Pread, SizeWith)]
     |                           ^^^^^ the trait `TryFromCtx<'_, _>` is not implemented for `[u16; 40]`
...
1331 | / multi_structs! {
1332 | |     /// Miscellaneous process information
1333 | |     ///
1334 | |     /// This struct matches the [Microsoft struct][msdn] of the same name.
...    |
1376 | |     }
1377 | | }
     | |_- in this macro invocation
     |
     = help: the following implementations were found:
               <&'a [u8] as TryFromCtx<'a, usize>>
     = note: this error originates in the derive macro `Pread` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `size_with` found for array `[u16; 260]` in the current scope
    --> /home/emz/.cargo/git/checkouts/rust-minidump-836c4af26fbfd1e5/4d95707/minidump-common/src/format.rs:1322:32
     |
1322 |           #[derive(Clone, Pread, SizeWith)]
     |                                  ^^^^^^^^ function or associated item not found in `[u16; 260]`
...
1331 | / multi_structs! {
1332 | |     /// Miscellaneous process information
1333 | |     ///
1334 | |     /// This struct matches the [Microsoft struct][msdn] of the same name.
...    |
1376 | |     }
1377 | | }
     | |_- in this macro invocation
     |
     = note: this error originates in the derive macro `SizeWith` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `size_with` found for array `[u16; 40]` in the current scope
    --> /home/emz/.cargo/git/checkouts/rust-minidump-836c4af26fbfd1e5/4d95707/minidump-common/src/format.rs:1322:32
     |
1322 |           #[derive(Clone, Pread, SizeWith)]
     |                                  ^^^^^^^^ function or associated item not found in `[u16; 40]`
...
1331 | / multi_structs! {
1332 | |     /// Miscellaneous process information
1333 | |     ///
1334 | |     /// This struct matches the [Microsoft struct][msdn] of the same name.
...    |
1376 | |     }
1377 | | }
     | |_- in this macro invocation
     |
     = note: this error originates in the derive macro `SizeWith` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `minidump-common`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
error: Setup script exited with 101
````

**Comment by BYK** [comment]
>do I have to run snuba version in sync with sentry (in general) ? I've tried to build 21.8.0 and it builds just fine.

Yes but I think in this case you can probably go past it. We can only guarantee they would work without issues if they have matching versions as in time certain things go out of sync.

>furthermore, do I even need snuba/relay to follow the upgrade path - do I understand correctly that I only need sentry (which I have built) to run it's 21.5.0 migration sequence (sentry upgrade launch) against clickhouse/postgresql, then proceed to the recent 21.x release ?

Again, this is not always the case but for this specific instance, you can only get by by upgrading Sentry first and then using the sync versions once the DB migrations are completed.

Any reason why you are not using the Docker images we build?

**Comment by drook** [comment]
Thanks a lot for your answers. I was able to upgrade to 21.6.3.

As about why I don't use the provided docker images, it's because
- using standalone docker starting from compose run looks jiust unfinished, and I have a relatively loaded instance. so it looks logical either to port it to k8s (which I'm not sure I'm capable of) or running is in a standalone environment.
- running containerless gives you a flexibility: we use dedicated databases - both CH and PgSQL
- when you started disassembling docker-compose bundle, all you have to do is to finish it, otherwise just don't start with it. :)

I also have a docker-compose instance running on another site, in the end I have to deal with same issues, leaving the build issues out of the scope, but the letter doesn't bother me much, so....

**Comment by BYK** [comment]
>using standalone docker starting from compose run looks jiust unfinished, and I have a relatively loaded instance. so it looks logical either to port it to k8s (which I'm not sure I'm capable of) or running is in a standalone environment.

Why do you think it is unfinished? That setup is being used by quite a lot of people and it is also the basis of this Helm chart if you really want to use k8s: https://github.com/sentry-kubernetes/charts

>running containerless gives you a flexibility: we use dedicated databases - both CH and PgSQL

You can still swap in containerless CH and Pg by changing some configs and bypassing related parts in the scripts.

>when you started disassembling docker-compose bundle, all you have to do is to finish it, otherwise just don't start with it. :)

Can you clarify this part? I don't quite get what you are saying.

Anyway, since this issue is resolved and we cannot go back in time to fix 21.5 builds, I am inclined to close this issue. Do you think there's something we can/should do regarding this issue?
